pwndbg/commands/leakfind.py

```diff
@@ -0,0 +1,155 @@
+parser.add_argument('--negative_offset',nargs="?",default=0x0,help="Max negative offset to search before an address when looking for a leak.")
+@pwndbg.commands.ArgparsedCommand(parser)
+@pwndbg.commands.OnlyWhenRunning
+def leakfind(address=-1,page_name=None,max_offset=0x40,max_depth=0x4,stride=0x1,negative_offset=0x0):
+    if address == -1:
+        print("No starting address provided. Please run leakfind -h for more information.")
+        return
+
+    foundPages = getPage(address)
+
+    if not foundPages:
+        print("Starting address is not mapped. Please run leakfind -h for more information.")
+        return
+
+    if not pwndbg.memory.peek(address):
+        print("Unable to read from starting address. Please run leakfind -h for more information.")
+        return
+
+    max_depth=int(max_depth)
+    #Just warn the user that a large depth might be slow. Probably worth checking offset^depth < threshold. Do this when more benchmarking is established.
+    if max_depth > 8:
+        print("leakfind may take a while to run on larger depths.")
+    
+    stride = int(stride)
+    address=int(address)
+    max_offset = int(max_offset)
+    negative_offset=int(negative_offset)
+    
+    #The below map stores a map of child address->(parent_address,parent_start_address)
+    #In the above tuple, parent_address is the exact address with a pointer to the child adddress.
+    #parent_start_address is an address that a previous address pointed to.
+    #We need to store both so that we can nicely create our leak chain.
+    visitedMap = {}
+    visitedSet = set()
+    visitedSet.add(int(address))
+    addressQueue = Queue()
+    addressQueue.put(int(address))
+    depth = 0
+    timeToDepthIncrease=0
+
+    #Run a bfs
+    #TODO look into performance gain from checking if an address is mapped before calling pwndbg.memory.pvoid()
+    #TODO also check using pwndbg.memory.read for possible performance boosts.
+    while addressQueue.qsize() > 0 and depth < max_depth:
+        if timeToDepthIncrease == 0:
+            depth=depth+1
+            timeToDepthIncrease=addressQueue.qsize()
+        cur_start_addr = addressQueue.get()
+        timeToDepthIncrease-=1
+        for cur_addr in range(cur_start_addr-negative_offset,cur_start_addr+max_offset,stride):
+            try:
+                result = int(pwndbg.memory.pvoid(cur_addr))
+                if result in visitedMap:
+                    continue
+                if result in visitedSet:
+                    continue
+                visitedMap[result]=(cur_addr,cur_start_addr) #map is of form child->(parent,parent_start)
+                addressQueue.put(result)
+                visitedSet.add(result)
+            except gdb.error:
+                #That means the memory was unreadable. Just skip it if we can't read it. 
```

Comment:
Is that correct? Can memory be executable but not readable? Would we miss an interesting pointer in such case?

Reply:
I'm not all that familiar with gdb's python API. I meant to say that the memory is unreadable by gdb i.e. unmapped, rather than saying that the protections of the memory didn't include readable. 
The line that would throw an error is the call to pwndbg.memory.pvoid. 
Is the correct interpretation that the memory is unmapped rather than unreadable?

Reply:
Ah, that makes sense.

> Is the correct interpretation that the memory is unmapped rather than unreadable?
I'm not 100% sure atm but I think yes